include/irr/asset/IMeshPacker.h

```diff
@@ -21,13 +18,96 @@ template <typename MeshBufferType>
+        :m_maxIndexCountPerMDIData(maxIndexCountPerMDIData), m_drawCount(0)
+    {
+        m_outVtxInputParams.enabledAttribFlags  = preDefinedLayout.enabledAttribFlags;
+        m_outVtxInputParams.enabledBindingFlags = preDefinedLayout.enabledBindingFlags;
+
+        memcpy(m_outVtxInputParams.attributes, preDefinedLayout.attributes, sizeof(m_outVtxInputParams.attributes));
+
+        m_outMDIData = core::make_smart_refctd_ptr<ICPUBuffer>(allocParams.MDIDataSizeInBytes);
+
+        //TODO: init allocators 
+        //m_alctrReservedSpace[0] = _IRR_ALIGNED_MALLOC(core::GeneralpurposeAddressAllocator::reserved_size(/**/), _IRR_SIMD_ALIGNMENT);
+        //m_vtxBuffAlctr = core::GeneralpurposeAddressAllocator(m_alctrReservedSpace[0], 0u, 0u, 0u, allocParams.vertexBuffSupportedSizeInBytes / allocParams.vertexbufferMinAllocSize, allocParams.vertexbufferMinAllocSize);
+
+        //m_alctrReservedSpace[1] = _IRR_ALIGNED_MALLOC(core::GeneralpurposeAddressAllocator::reserved_size(/**/), _IRR_SIMD_ALIGNMENT);
+        //m_idxBuffAlctr = core::GeneralpurposeAddressAllocator(m_alctrReservedSpace[1], 0u, 0u, 0u, allocParams.indexBuffSupportedSizeInBytes / allocParams.indexBufferMinAllocSize, allocParams.indexBufferMinAllocSize);
+
+        //m_alctrReservedSpace[2] = _IRR_ALIGNED_MALLOC(core::GeneralpurposeAddressAllocator::reserved_size(/**/), _IRR_SIMD_ALIGNMENT);
+        //m_perInsVtxBuffAlctr = core::GeneralpurposeAddressAllocator(m_alctrReservedSpace[2], 0u, 0u, 0u, allocParams.perInstanceVertexBuffSupportedSizeInBytes / allocParams.perInstanceVertexbufferMinAllocSize, allocParams.perInstanceVertexbufferMinAllocSize);
+        
+
+
+        //init MDI data allocator
+        m_alctrReservedSpace[3] = static_cast<uint8_t*>(_IRR_ALIGNED_MALLOC(core::GeneralpurposeAddressAllocator<size_t>::reserved_size(alignof(uint32_t), allocParams.MDIDataSizeInBytes / sizeof(DrawElementsIndirectCommand_t), 1u), _IRR_SIMD_ALIGNMENT));
+        m_MDIDataAlctr = core::GeneralpurposeAddressAllocator<size_t>(m_alctrReservedSpace[3], 0u, 0u, alignof(uint32_t), allocParams.MDIDataSizeInBytes / sizeof(DrawElementsIndirectCommand_t), /*i git*/1u);
+
+        //1 attrib enabled == 1 binding
+        for (uint16_t attrBit = 0x0001, location = 0; location < 16; attrBit <<= 1, location++)
+        {
+            if (m_outVtxInputParams.enabledAttribFlags & attrBit)
+            {
+                m_outVtxInputParams.bindings[location].stride = getTexelOrBlockBytesize(static_cast<E_FORMAT>(m_outVtxInputParams.attributes[location].format));
+                m_outVtxInputParams.bindings[location].inputRate = preDefinedLayout.bindings[preDefinedLayout.attributes[location].binding].inputRate;
+            }
+        }
+    }
+
+    virtual std::optional<PackedMeshBufferData> alloc(const MeshBufferType* meshBuffer) = 0;
```

Comment:
don't use `std::optional` on frontend API, add an `isValid` method to `PackedMeshBufferData`

Reply:
> don't use std::optional on frontend API
why?

Reply:
why?

Reply:
it makes the code less-readable, wastes memory (add extra members)

these heavy duty functions will be dealing with thousands of meshes, so we want the inputs and outputs to be simple structs which are PoD (Plain Old Data)